docs/examples/test_examples.py

```diff
@@ -0,0 +1,268 @@
+from nameko.testing.services import entrypoint_waiter
+from nameko.testing.websocket import make_virtual_socket
+
+
+@pytest.yield_fixture
+def container_factory(rabbit_config):
+
+    all_containers = []
+
+    def make_container(service_cls, config, worker_ctx_cls=None):
+        container = ServiceContainer(service_cls, config, worker_ctx_cls)
+        all_containers.append(container)
+        return container
+
+    yield make_container
+
+    for c in all_containers:
+        try:
+            c.stop()
+        except:
+            pass
+
+
+@pytest.yield_fixture()
+def websocket():
+    active_sockets = []
+
+    def socket_creator():
+        ws_app, wait_for_sock = make_virtual_socket(
+            '127.0.0.1', 8000)
+        gr = eventlet.spawn(ws_app.run_forever)
+        active_sockets.append((gr, ws_app))
+        return wait_for_sock()
+
+    try:
+        yield socket_creator
+    finally:
+        for gr, ws_app in active_sockets:
+            try:
+                ws_app.close()
+            except Exception:
+                pass
+            gr.kill()
+
+
+class TestHttp(object):
+
+    base_url = "http://localhost:8000"
+
+    def test_http(self, container_factory, rabbit_config):
+
+        from http import HttpService
+
+        container = container_factory(HttpService, rabbit_config)
+        container.start()
+
+        res = requests.get(self.base_url + "/get/42")
+        assert res.status_code == 200
+        assert res.text == '{"value": 42}'
+
+        res = requests.post(self.base_url + "/post", data="Hello")
+        assert res.status_code == 200
+        assert res.text == 'received: Hello'
+
+    def test_advanced(self, container_factory, rabbit_config):
+
+        from advanced_http import Service
+
+        container = container_factory(Service, rabbit_config)
+        container.start()
+
+        res = requests.get(self.base_url + "/privileged")
+        assert res.status_code == 403
+        assert res.text == 'Forbidden'
+
+        res = requests.get(self.base_url + "/headers")
+        assert res.status_code == 201
+        assert res.headers['location'] == 'https://www.example.com/widget/1'
+
+        res = requests.get(self.base_url + "/custom")
+        assert res.status_code == 200
+        assert res.text == 'payload'
+
+
+class TestEvents(object):
+
+    def test_events(self, container_factory, rabbit_config):
+
+        from events import ServiceA, ServiceB
+
+        container_a = container_factory(ServiceA, rabbit_config)
+        container_b = container_factory(ServiceB, rabbit_config)
+        container_a.start()
+        container_b.start()
+
+        with ServiceRpcProxy('service_a', rabbit_config) as service_a_rpc:
+
+            with patch.object(ServiceB, 'handle_event') as handle_event:
+
+                with entrypoint_waiter(container_b, 'handle_event'):
+                    service_a_rpc.dispatching_method()
+                assert handle_event.call_args_list == [call("payload")]
```

Comment:
is it confusing that the variable `payload` ends up containing the string `"payload"`?